wire-ios-data-model/Source/Model/UserClient/UserClient.swift

```diff
@@ -503,37 +500,98 @@ public extension UserClient {
-                // TODO: [John] use flag here
-                syncSelfClient.keysStore.encryptionContext.perform({ (sessionsDirectory) in
-                    syncClient.fingerprint = sessionsDirectory.localFingerprint
-                    syncMOC.saveOrRollback()
-                })
-            }
-            else {
+                syncClient.fingerprint = syncClient.localFingerprint
+                syncMOC.saveOrRollback()
+            } else {
                 if !syncClient.hasSessionWithSelfClient {
                     syncSelfClient.missesClient(syncClient)
                     syncSelfClient.setLocallyModifiedKeys(Set(arrayLiteral: ZMUserClientMissingKey))
                     syncMOC.saveOrRollback()
                 }
                 else {
-                    // TODO: [John] use flag here
-                    syncSelfClient.keysStore.encryptionContext.perform({ (sessionsDirectory) in
-                        syncClient.fingerprint = sessionsDirectory.fingerprint(for: sessionIdentifier)
-                        if syncClient.fingerprint == nil {
-                            zmLog.error("Cannot fetch fingerprint for client \(syncClient.sessionIdentifier!)")
-                        } else {
-                            syncMOC.saveOrRollback()
-                        }
-                    })
+                    syncClient.fingerprint = syncClient.remoteFingerprint
+                    guard syncClient.fingerprint != nil else {
+                        zmLog.error("Cannot fetch fingerprint for client \(syncClient.sessionIdentifier!)")
+                        return
+                    }
+                    syncMOC.saveOrRollback()
                 }
             }
         })
     }
+
+}
+
+// MARK: - Fetch fingerprint
+
+extension UserClient {
+
+    private var remoteFingerprint: Data? {
+        guard let proteusProvider = self.proteusProvider,
+              let sessionIdentifier = self.sessionIdentifier else {
+                  return nil
+              }
+
+        var fingerprintData: Data? = nil
+
+        proteusProvider.perform(
+            withProteusService: { proteusService in
+                guard let proteusSessionID = self.proteusSessionID else {
+                    return
+                }
+
+                do {
+                    let fingerprint = try proteusService.remoteFingerprint(forSession: proteusSessionID)
+                    fingerprintData = fingerprint.utf8Data
+                } catch {
+                    zmLog.error("Cannot fetch remote fingerprint for \(self)")
+                }
+            },
+
+            withKeyStore: { keyStore in
+                keyStore.encryptionContext.perform({ (sessionsDirectory) in
+                    fingerprintData = sessionsDirectory.fingerprint(for: sessionIdentifier)
+                    print(fingerprintData)
+                })
+            }
+        )
+
+        return fingerprintData
+    }
+
+    private var localFingerprint: Data? {
```

Comment:
I think this should be a method rather than a computed property because it involves a bit of work which is not apparent if it's exposed as a property.

Reply:
Done